### PR TITLE
SQL Connection implementation

### PR DESCRIPTION
**Pull request by fedefernandez**
This PR brings the SQL Connection implementation

Also includes a connection url parser to create connections

The file `SQLDroidPreparedStatement.java` should be ignored because will be implemented in Scala

Please @juanpedromoreno could you review? Thanks

### CODE REVIEW

File: src/test/scala/com/fortysevendeg/android/sqlite/SQLDroidConnectionSpec.scala [hunk 1]
```diff
@@ -0,0 +1,638 @@
...
+
+    val sqlDroid: SQLDroidConnection
+
+  }
+
+  trait WithSomeConnection extends SQLDroidConnectionScope {
+
+    var lastPreparedSql: String = ""
+
+    var lastPreparedColumn: Option[String] = None
```
Comment (47degci) [inline src/test/scala/com/fortysevendeg/android/sqlite/SQLDroidConnectionSpec.scala]: Codacy Issue found: [Avoid mutable fields](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/809217350/issues/source?bid=2442814&fileBranchId=2738989#l47)

File: src/test/scala/com/fortysevendeg/android/sqlite/SQLDroidConnectionSpec.scala [hunk 2]
```diff
@@ -0,0 +1,638 @@
...
+
+    database.database returns sqliteDatabase
+
+    val sqlDroid: SQLDroidConnection
+
+  }
+
+  trait WithSomeConnection extends SQLDroidConnectionScope {
+
+    var lastPreparedSql: String = ""
```
Comment (47degci) [inline src/test/scala/com/fortysevendeg/android/sqlite/SQLDroidConnectionSpec.scala]: Codacy Issue found: [Avoid mutable fields](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/809217350/issues/source?bid=2442814&fileBranchId=2738989#l45)

File: src/main/scala/com/fortysevendeg/android/sqlite/util/ConnectionStringParser.scala [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  private[this] def readParams(firstParam: String, lastParams: String): Map[String, String] =
+    (Option(firstParam), Option(lastParams)) match {
+      case (Some(f), Some(t)) if f.length > 0 => paramsToMap(f) ++ paramsToMap(cleanParams(t): _*)
+      case (Some(f), None) if f.length > 0 => paramsToMap(f)
+      case _ => Map.empty
+    }
+
+  private[this] def paramsToMap(params: String*): Map[String, String] =
+    Map(params map { p =>
+      val Array(name, value) = p.split("=")
```
Comment (47degci) [inline src/main/scala/com/fortysevendeg/android/sqlite/util/ConnectionStringParser.scala]: Codacy Issue found: [Field name does not match the regular expression '^[a-z][A-Za-z0-9]*$'](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/809217347/issues/source?bid=2442814&fileBranchId=2738989#l32)

File: src/test/scala/com/fortysevendeg/android/sqlite/SQLDroidConnectionSpec.scala [hunk 4]
```diff
@@ -0,0 +1,638 @@
...
+      }
+    }
+
+  }
+
+  trait WithSomeConnectionAutoCommit extends SQLDroidConnectionScope {
+
+    var lastPreparedSql: String = ""
+
+    var lastPreparedColumn: Option[String] = None
```
Comment (47degci) [inline src/test/scala/com/fortysevendeg/android/sqlite/SQLDroidConnectionSpec.scala]: Codacy Issue found: [Avoid mutable fields](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/809217350/issues/source?bid=2442814&fileBranchId=2738989#l70)

File: src/main/scala/com/fortysevendeg/android/sqlite/SQLDroidConnection.scala [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+
+  val closingErrorMessage = "Error closing database"
+
+  val autoCommitErrorMessage = "Database is in auto-commit mode"
+
+  val alreadyClosedErrorMessage = "Database connection closed"
+
+  var sqliteDatabase: Option[SQLDroidDatabase] = createDatabase()
+
+  var autoCommit: Boolean = defaultAutoCommit()
```
Comment (47degci) [inline src/main/scala/com/fortysevendeg/android/sqlite/SQLDroidConnection.scala]: Codacy Issue found: [Avoid mutable fields](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/809217331/issues/source?bid=2442814&fileBranchId=2738989#l45)

File: src/test/scala/com/fortysevendeg/android/sqlite/SQLDroidConnectionSpec.scala [hunk 6]
```diff
@@ -0,0 +1,638 @@
...
+      }
+    }
+
+  }
+
+  trait WithNoneConnection extends SQLDroidConnectionScope {
+
+    var lastPreparedSql: String = ""
+
+    var lastPreparedColumn: Option[String] = None
```
Comment (47degci) [inline src/test/scala/com/fortysevendeg/android/sqlite/SQLDroidConnectionSpec.scala]: Codacy Issue found: [Avoid mutable fields](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/809217350/issues/source?bid=2442814&fileBranchId=2738989#l95)

File: src/main/scala/com/fortysevendeg/android/sqlite/SQLDroidConnection.scala [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+
+  val rollbackSql = "rollback;"
+
+  val closingErrorMessage = "Error closing database"
+
+  val autoCommitErrorMessage = "Database is in auto-commit mode"
+
+  val alreadyClosedErrorMessage = "Database connection closed"
+
+  var sqliteDatabase: Option[SQLDroidDatabase] = createDatabase()
```
Comment (47degci) [inline src/main/scala/com/fortysevendeg/android/sqlite/SQLDroidConnection.scala]: Codacy Issue found: [Avoid mutable fields](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/809217331/issues/source?bid=2442814&fileBranchId=2738989#l43)

File: src/test/scala/com/fortysevendeg/android/sqlite/SQLDroidConnectionSpec.scala [hunk 8]
```diff
@@ -0,0 +1,638 @@
...
+        javaNull
+
+      }
+    }
+
+  }
+
+  trait WithSomeConnectionAutoCommit extends SQLDroidConnectionScope {
+
+    var lastPreparedSql: String = ""
```
Comment (47degci) [inline src/test/scala/com/fortysevendeg/android/sqlite/SQLDroidConnectionSpec.scala]: Codacy Issue found: [Avoid mutable fields](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/809217350/issues/source?bid=2442814&fileBranchId=2738989#l68)

File: src/main/scala/com/fortysevendeg/android/sqlite/SQLDroidConnection.scala [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+
+import java.sql._
+import java.util.Properties
+import java.util.concurrent.Executor
+
+import com.fortysevendeg.android.sqlite.logging.{AndroidLogWrapper, LogWrapper}
+import com.fortysevendeg.android.sqlite.metadata.SQLDroidDatabaseMetaData
+import org.sqldroid.{SQLDroidPreparedStatement, SQLDroidStatement}
+
+class SQLDroidConnection(
```
Comment (47degci) [inline src/main/scala/com/fortysevendeg/android/sqlite/SQLDroidConnection.scala]: Codacy Issue found: [Avoid using finalize method.](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/809217331/issues/source?bid=2442814&fileBranchId=2738989#l11)

Comment (47degci) [inline src/main/scala/com/fortysevendeg/android/sqlite/SQLDroidConnection.scala]: Codacy Issue found: [Number of methods in class exceeds 30](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/809217331/issues/source?bid=2442814&fileBranchId=2738989#l11)

File: src/main/scala/com/fortysevendeg/android/sqlite/SQLDroidDriver.scala [hunk 10]
```diff
@@ -25,7 +26,54 @@ class SQLDroidDriver extends Driver {
...
+
+  private[this] def readTimeOut(params: Map[String, String]): Option[Long] =
+    readLongParam("timeout", params)
+
+  private[this] def readRetry(params: Map[String, String]): Option[Int] =
+    readLongParam("retry", params) map (_.toInt)
+
+  private[this] def readLongParam(name: String, params: Map[String, String]): Option[Long] =
+    params.get(name) flatMap { value =>
+      Try(value.toInt) match {
```
Comment (juanpedromoreno) [inline src/main/scala/com/fortysevendeg/android/sqlite/SQLDroidDriver.scala]: Try(value.toInt).toOption.

And why not `value.toLong` if you want a `Long`?

File: src/main/scala/com/fortysevendeg/android/sqlite/SQLDroidDriver.scala [hunk 11]
```diff
@@ -25,7 +26,54 @@ class SQLDroidDriver extends Driver {
...
+
+  private[this] def completeFlags(databaseFlags: String, androidFlags: String): Int =
+    (Option(databaseFlags), Option(androidFlags)) match {
+      case (Some(f), _) => parseInt(f, flags)
+      case (None, Some(f)) => flags | parseInt(f, 0)
+      case _ => flags
+    }
+
+  private[this] def parseInt(value: String, defaultValue: Int) =
+    Try(value.toInt) match {
```
Comment (juanpedromoreno) [inline src/main/scala/com/fortysevendeg/android/sqlite/SQLDroidDriver.scala]: `Try(value.toInt).toOption getOrElse defaultValue`?

File: test/org/sqldroid/DriverUnitTest.java [hunk 12]
```diff
@@ -674,7 +674,7 @@ public void testAutoCommit() throws Exception {
         String jdbcURL = JDBC_URL_PREFIX + dbFile;               
  
         Properties removeLocale = new Properties();
-        removeLocale.put(SQLDroidDriver.ADDITONAL_DATABASE_FLAGS, android.database.sqlite.SQLiteDatabase.NO_LOCALIZED_COLLATORS);
+        removeLocale.put("AdditionalDatabaseFlags", android.database.sqlite.SQLiteDatabase.NO_LOCALIZED_COLLATORS);
```
Comment (juanpedromoreno) [inline test/org/sqldroid/DriverUnitTest.java:677]: does it make sense if we put this constant into a val?

File: src/test/scala/com/fortysevendeg/android/sqlite/util/ConnectionStringParserSpec.scala [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+
+    "return None when the url is doesn't have a valid prefix" in
+      new ConnectionStringParserScope {
+        parseConnectionString(urlWithInvalidPrefix) must beNone
+      }
+
+    "return a memory name and an empty map params when passing a memory url" in
+      new ConnectionStringParserScope {
+        parseConnectionString(urlInMemory) must beSome[ConnectionValues].which { c =>
+          c shouldEqual ConnectionValues(":memory", Map.empty)
```
Comment (juanpedromoreno) [inline src/test/scala/com/fortysevendeg/android/sqlite/util/ConnectionStringParserSpec.scala]: does it make sense if we put this `:memory` into a val?

File: src/main/scala/com/fortysevendeg/android/sqlite/SQLDroidConnection.scala [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+  databaseName: String,
+  timeout: Long = 0,
+  retryInterval: Int = 50,
+  flags: Int = 0,
+  val logWrapper: LogWrapper = new AndroidLogWrapper()
+  ) extends Connection with WrapperNotSupported {
+
+  // TODO - Method for compatibility with java classes. Remove after migration of this classes
```
Comment (juanpedromoreno) [inline src/main/scala/com/fortysevendeg/android/sqlite/SQLDroidConnection.scala:19]: just a question, have been these classes already migrated? It's just to be sure we cannot remove this method yet.

**Comment by juanpedromoreno** [comment]
Minor comments added. LGTM! Good job.

File: test/org/sqldroid/DriverUnitTest.java (hunk 12 above)
Comment (fedefernandez) [inline test/org/sqldroid/DriverUnitTest.java:677]: This file will be removed in a future PR. Didn't remove yet to consult previous tests

File: src/main/scala/com/fortysevendeg/android/sqlite/SQLDroidConnection.scala (hunk 14 above)
Comment (fedefernandez) [inline src/main/scala/com/fortysevendeg/android/sqlite/SQLDroidConnection.scala:19]: Not yet, the dependency is with the the last two migrations (`Statement` and `PreparedStatement`)

File: src/test/scala/com/fortysevendeg/android/sqlite/util/ConnectionStringParserSpec.scala (hunk 13 above)
Comment (fedefernandez) [inline src/test/scala/com/fortysevendeg/android/sqlite/util/ConnectionStringParserSpec.scala]: Yes :smile:

File: src/main/scala/com/fortysevendeg/android/sqlite/SQLDroidDriver.scala (hunk 11 above)
Comment (fedefernandez) [inline src/main/scala/com/fortysevendeg/android/sqlite/SQLDroidDriver.scala]: :+1:

File: src/main/scala/com/fortysevendeg/android/sqlite/SQLDroidDriver.scala (hunk 10 above)
Comment (fedefernandez) [inline src/main/scala/com/fortysevendeg/android/sqlite/SQLDroidDriver.scala]: Good catch

File: src/main/scala/com/fortysevendeg/android/sqlite/logging/LogWrapper.scala [hunk 15]
```diff
@@ -14,7 +16,17 @@ trait LogWrapper {
...
   def e(message: String, maybeThrowable: Option[Throwable] = None)
 
-  def notImplemented[T](result: T): T
+  def notImplemented[T](result: T): T = {
+    val message = "Method not implemented" :: (stackTraces map (s => s" at ${methodName(s)}(${fileName(s)})")).toList
+    w(message.mkString("\n"))
+    result
+  }
+
+  def logOnError[T](f: => T, msg: String = "") =
```
Comment (47degci) [inline src/main/scala/com/fortysevendeg/android/sqlite/logging/LogWrapper.scala:25]: Codacy Issue found: [Call by name parameter should be in last position](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/809264718/issues/source?bid=2442814&fileBranchId=2738989#l25)

File: src/test/scala/com/fortysevendeg/android/sqlite/SQLDroidConnectionSpec.scala (hunk 1 above)
Comment (47degci) [inline src/test/scala/com/fortysevendeg/android/sqlite/SQLDroidConnectionSpec.scala:47]: Codacy Issue found: [Avoid mutable fields](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/809264730/issues/source?bid=2442814&fileBranchId=2738989#l47)

File: src/test/scala/com/fortysevendeg/android/sqlite/SQLDroidConnectionSpec.scala (hunk 2 above)
Comment (47degci) [inline src/test/scala/com/fortysevendeg/android/sqlite/SQLDroidConnectionSpec.scala:45]: Codacy Issue found: [Avoid mutable fields](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/809264730/issues/source?bid=2442814&fileBranchId=2738989#l45)

File: src/main/scala/com/fortysevendeg/android/sqlite/util/ConnectionStringParser.scala (hunk 3 above)
Comment (47degci) [inline src/main/scala/com/fortysevendeg/android/sqlite/util/ConnectionStringParser.scala:32]: Codacy Issue found: [Field name does not match the regular expression '^[a-z][A-Za-z0-9]*$'](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/809264727/issues/source?bid=2442814&fileBranchId=2738989#l32)

File: src/test/scala/com/fortysevendeg/android/sqlite/SQLDroidConnectionSpec.scala (hunk 4 above)
Comment (47degci) [inline src/test/scala/com/fortysevendeg/android/sqlite/SQLDroidConnectionSpec.scala:70]: Codacy Issue found: [Avoid mutable fields](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/809264730/issues/source?bid=2442814&fileBranchId=2738989#l70)

File: src/main/scala/com/fortysevendeg/android/sqlite/SQLDroidConnection.scala (hunk 5 above)
Comment (47degci) [inline src/main/scala/com/fortysevendeg/android/sqlite/SQLDroidConnection.scala:45]: Codacy Issue found: [Avoid mutable fields](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/809264711/issues/source?bid=2442814&fileBranchId=2738989#l45)

File: src/test/scala/com/fortysevendeg/android/sqlite/SQLDroidConnectionSpec.scala (hunk 6 above)
Comment (47degci) [inline src/test/scala/com/fortysevendeg/android/sqlite/SQLDroidConnectionSpec.scala:95]: Codacy Issue found: [Avoid mutable fields](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/809264730/issues/source?bid=2442814&fileBranchId=2738989#l95)

File: src/main/scala/com/fortysevendeg/android/sqlite/SQLDroidConnection.scala (hunk 7 above)
Comment (47degci) [inline src/main/scala/com/fortysevendeg/android/sqlite/SQLDroidConnection.scala:43]: Codacy Issue found: [Avoid mutable fields](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/809264711/issues/source?bid=2442814&fileBranchId=2738989#l43)

File: src/test/scala/com/fortysevendeg/android/sqlite/SQLDroidConnectionSpec.scala (hunk 8 above)
Comment (47degci) [inline src/test/scala/com/fortysevendeg/android/sqlite/SQLDroidConnectionSpec.scala:68]: Codacy Issue found: [Avoid mutable fields](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/809264730/issues/source?bid=2442814&fileBranchId=2738989#l68)

File: src/main/scala/com/fortysevendeg/android/sqlite/SQLDroidConnection.scala (hunk 9 above)
Comment (47degci) [inline src/main/scala/com/fortysevendeg/android/sqlite/SQLDroidConnection.scala:11]: Codacy Issue found: [Avoid using finalize method.](https://www.codacy.com/app/47deg/scala-sqlite-droid/file/809264711/issues/source?bid=2442814&fileBranchId=2738989#l11)

**Pull request merged**